lms/views/lti/basic_launch.py

```diff
@@ -101,12 +95,12 @@ def db_configured_launch(self):
         # The ``db_configured=True`` view predicate ensures that this view
         # won't be called if there isn't a matching document_url in the DB. So
         # here we can safely assume that the document_url exists.
-        document_url = self.assignment_service.get_assignment(
-            self.context.lti_params["tool_consumer_instance_guid"],
-            self.context.resource_link_id,
-        ).document_url
-
-        return self._do_launch(document_url=document_url)
+        return self._do_launch(
+            document_url=self.assignment_service.get_assignment(
+                self.context.lti_params["tool_consumer_instance_guid"],
+                self.context.resource_link_id,
+            ).document_url
+        )
 
     @view_config(url_configured=True, schema=URLConfiguredBasicLTILaunchSchema)
     def url_configured_launch(self):
```

Comment:
There's a subtle going on here left from the last attempt:

- `db_configured_launch` has both `db_configured=True` and `url_configured=False`
- `url_configured_launch` has just `url_configured=True`


So before this change `db_configured` and `url_configured` never were true at the same time but now the can and `url_configured_launch` is taking precedence.

I think every test would pass if you remove the `url_configured=False` from `db_configured_launch` but assignment editing will stop working on Canvas.

We talk about how the predicates hide this things so that's something to to have in mind for an eventual refactor. 

Looks like a for a lot of these, what's different about them is how we construct/where we pick from the `document_url` maybe that could help in predicate-less refactor.

Reply:
Yeah... they are hugely confusing. I think this is a "hiding" rather than "revealing" abstraction. The details are tucked away in such a way as it's difficult to think about them.

I don't believe this PR changes anything about these though?

Reply:
No, not related to this PR

I was trying to find some edge  case where having the row in the DB would confuse the existing code but I haven't found any.

Reply:
I was working through the logic of all of this, and I think there are basically two main cases for the configured launches:

 * Those that read from params to get the URL
 * Those that read from the assignment (course copied + configured)
 
In the case of the param based ones, the presence of the assignment is irrelevant. It's upsert either way. The other types do require it to exist. 

So I think the work this might be trying to do is:

 * `url_configured_launch` 
    * `url_configured=True` - This is correct, we don't care about db configured or not
 * `db_configured_launch`
    * `db_configured=True` - This is correct, it's required
    * `url_configured=False` - This feels like an attempt to avoid precidence issues. This is "true", but unnecessary if url configured kicks in first. It's also not complete, as really this should be "course_copied=False" as well etc. So I'm not sure if this is required or not